pkg/monitor/azure/nsg/nsg.go

```diff
@@ -0,0 +1,168 @@
+	MetricNSGMonitoringTimedOut    = "monitor.preconfigurednsg.monitoringtimedout"
+)
+
+var expandNSG = "NetworkSecurityGroup"
+
+// NSGMonitor is responsible for performing NSG rule validations when preconfiguredNSG is enabled
+type NSGMonitor struct {
+	log     *logrus.Entry
+	emitter metrics.Emitter
+	oc      *api.OpenShiftCluster
+
+	wg *sync.WaitGroup
+
+	subnetClient *armnetwork.SubnetsClient
+	dims         map[string]string
+}
+
+func NewNSGMonitor(log *logrus.Entry, oc *api.OpenShiftCluster, subscriptionID string, subnetClient *armnetwork.SubnetsClient, emitter metrics.Emitter, wg *sync.WaitGroup) *NSGMonitor {
+	return &NSGMonitor{
+		log:     log,
+		emitter: emitter,
+		oc:      oc,
+
+		subnetClient: subnetClient,
+		wg:           wg,
+
+		dims: map[string]string{
+			dimension.ResourceID:     oc.ID,
+			dimension.SubscriptionID: subscriptionID,
+			dimension.Location:       oc.Location,
+		},
+	}
+}
+
+type subnetNSGConfig struct {
+	// subnet CIDR range
+	prefix []netip.Prefix
+	// The rules from the subnet NSG
+	nsg *armnetwork.SecurityGroup
+}
+
+func (n *NSGMonitor) toSubnetConfig(ctx context.Context, subnetID string) (subnetNSGConfig, error) {
+	r, err := arm.ParseResourceID(subnetID)
+	if err != nil {
+		return subnetNSGConfig{}, err
+	}
+
+	dims := map[string]string{
+		dimension.Subnet:        r.Name,
+		dimension.Vnet:          r.Parent.Name,
+		dimension.ResourceGroup: r.ResourceGroupName,
+	}
+
+	subnet, err := n.subnetClient.Get(ctx, r.ResourceGroupName, r.Parent.Name, r.Name, &armnetwork.SubnetsClientGetOptions{Expand: &expandNSG})
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr); respErr.StatusCode == http.StatusForbidden {
+			emitter.EmitGauge(n.emitter, MetricSubnetAccessForbidden, int64(1), n.dims, dims)
+		}
+		n.log.Errorf("error while getting subnet %s. %s", subnetID, err)
+		return subnetNSGConfig{}, err
+	}
+
+	var cidrs []string
+	if subnet.Properties.AddressPrefix != nil {
+		cidrs = append(cidrs, *subnet.Properties.AddressPrefix)
+
+	}
+	for _, sn := range subnet.Properties.AddressPrefixes {
+		cidrs = append(cidrs, *sn)
+	}
+	prefixes := toPrefixes(n.log, cidrs)
+	if len(prefixes) == 0 {
+		return subnetNSGConfig{}, errors.New("no valid subnet ranges found")
+	}
+	return subnetNSGConfig{prefixes, subnet.Properties.NetworkSecurityGroup}, nil
+}
+
+func (n *NSGMonitor) Monitor(ctx context.Context) []error {
+	defer n.wg.Done()
+	masterSubnet, err := n.toSubnetConfig(ctx, n.oc.Properties.MasterProfile.SubnetID)
+	if err != nil {
+		// FP has no access to the subnet
+		return []error{err}
+	}
+
+	// need this to get the right workerProfiles
+	workerProfiles, _ := api.GetEnrichedWorkerProfiles(n.oc.Properties)
+	workerSubnets := make([]subnetNSGConfig, 0, len(workerProfiles))
+	workerPrefixes := make([]netip.Prefix, 0, len(workerProfiles))
+	for _, wp := range workerProfiles {
+		s, err := n.toSubnetConfig(ctx, wp.SubnetID)
```

Comment:
I _think_ it's possible the subnetId can be empty should a customer misconfigure it.  We may want to have a check to skip should the subnet id be empty.  

We should test this to see if we can create a bad machineset with an invalid subnet and see if the enriched worker profiles has an empty subnet id.

Reply:
empty subnets should be ignored as the machinesets are invalid.

Reply:
Confirmed that the subnetID can be empty

```
            {
                "name": "created-with-an-invalid-subnet",
                "count": 1
            },
            {
                "name": "machine-set-name",
                "vmSize": "Standard_D4s_v3",
                "diskSizeGB": 128,
                "subnetId": "subnet-value",
                "count": 1,
                "encryptionAtHost": "Disabled"
            },
```

Reply:
add the logic to skip when a blank subnet is found.